Provision the Evergreen service layer

rtyler opened the pull request:
This is roughly following other examples for provisioning the evergreen
container

Fixes JENKINS-50985

@olblak I'm not entirely sure how to test this, but tossing this up here for now

Commit:
Provision the Evergreen service layer

This is roughly following other examples for provisioning the evergreen
container

Fixes JENKINS-50985

olblak commented:
I am happy to see this service running :) 
It would be nice to namespace this application like [lego](https://github.com/jenkins-infra/jenkins-infra/tree/staging/dist/profile/templates/kubernetes/resources/lego), I know that I almost never do that (and I should), but in this case the main advantage will be to provide a read only access to a small group of users to that namespace with RBAC.
RBAC is not supported with the current version but I plan to migrate the cluster to AKS which support version upgrade

olblak commented on dist/profile/manifests/kubernetes/resources/evergreen.pp:
```diff
@@ -0,0 +1,66 @@
+      resource => 'evergreen/service.yaml'
+    }
+
+    profile::kubernetes::apply { "evergreen/secret.yaml on ${context}":
```
I don't see any reference to a secret in the deployment so It sounds more like a wrong copy/paste

olblak replied:
Sorry so you need a secret but you don't provide the good parameter

olblak replied:
parameters should be:
```
parameters => {
  'postgres_url' => $postgres_url
}
```

olblak commented on dist/profile/templates/kubernetes/resources/evergreen/service.yaml.erb:
```diff
@@ -0,0 +1,17 @@
+    app: evergreen
+    type: evergreen
+spec:
+  sessionAffinity: ClientIP
```
do you need a session affinity?
The accountapp is the only application that need basically it ensure that a client wlll always be routed to the same pod

olblak commented on dist/profile/templates/kubernetes/resources/evergreen/secret.yaml.erb:
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: Secret 
+metadata: 
+    name: accountapp
```
I think you forgot this resource :P 
all current data aren't needed and you should have the postgres_key

olblak commented on dist/profile/templates/kubernetes/resources/evergreen/deployment.yaml.erb:
```diff
@@ -0,0 +1,37 @@
+metadata:
+  name: evergreen
+spec:
+  replicas: 1
```
Is it a stateless or statefull application? 
If it's a stateless app then I would increase the replicas to at least 3 in order to bring redundancy and the possibility to do  AB deployment.

olblak commented on dist/profile/manifests/kubernetes/resources/evergreen.pp:
```diff
@@ -0,0 +1,66 @@
+    Array $domain_alias = [],
+    String $image_tag = 'latest',
+    String $domain_name = 'evergreen.jenkins.io',
+    String $postgres_url = '',
```
Where is this postgres database? On azure?

olblak commented on dist/profile/manifests/kubernetes/resources/evergreen.pp:
```diff
@@ -0,0 +1,66 @@
+
+    profile::kubernetes::apply { "evergreen/deployment.yaml on ${context}":
+      context    => $context,
+      parameters => {
```
you are missing ```'image_tag'             => $image_tag``` in your parameter otherwise your deployment template will not not which tag to use

olblak commented:
I should document the way I test this kind of PRs :\

olblak commented on dist/profile/manifests/kubernetes/resources/evergreen.pp:
```diff
@@ -0,0 +1,66 @@
+class profile::kubernetes::resources::evergreen (
+    Array $clusters = $profile::kubernetes::params::clusters,
+    Array $domain_alias = [],
+    String $image_tag = 'latest',
```
If you plan to always us 'latest' tag, you'll need a way to tell kubernetes to restart running pods.
A trick would be  to add an information to the deployment resource that change every weeks/days/months
Something like 
metadata:
  date:  $(date "+%V")

olblak committed:
Add various evegreen deployment improvement

* Move evergreen application in it's own namespace
* Encode secret in base64
* Add image_tag variable to evergreen/deployment.yaml
* Increase number of evegreen pod
* Remove session affinity from ingress/service configuration
* Clean secret variable

Commit:
Add more secrets needed for the Evergreen application

Commit:
Add a specifically versioned image_tag for the evergreen backend

This is the latest with the necessary production changes from jenkins-infra/evergreen#160

Commit:
Add the evergreen CNAME for deploying the Jenkins Essentials backend
